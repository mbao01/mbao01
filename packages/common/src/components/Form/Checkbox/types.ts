import { type VariantProps } from "../../../libs";
import { getCheckboxClasses } from "./constants";

export type CheckboxProps = Omit<
  React.HTMLProps<HTMLInputElement>,
  "size" | "ref" | "type"
> &
  VariantProps<typeof getCheckboxClasses>;
