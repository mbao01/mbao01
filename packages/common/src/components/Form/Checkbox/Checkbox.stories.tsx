import type { Meta, StoryObj } from "@storybook/react";
import { useEffect, useRef } from "react";
import { FormControl } from "../components/FormControl";
import { Label } from "../components/Label";
import { Checkbox } from "./Checkbox";

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Checkbox,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof Checkbox>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {},
};

export const PrimaryCheckbox: Story = {
  args: {
    variant: "primary",
  },
};

export const TinyCheckbox: Story = {
  args: {
    size: "xs",
  },
};

export const DisabledCheckbox: Story = {
  args: {
    disabled: true,
  },
};

export const IntermediateCheckbox: Story = {
  render: (args) => {
    const ref = useRef<HTMLInputElement>(null);
    useEffect(() => {
      if (ref?.current) ref.current.indeterminate = true;
    }, []);

    return <Checkbox {...args} ref={ref} />;
  },
};

export const LabeledCheckbox: Story = {
  render: (args) => {
    return (
      <FormControl as="div" className="flex-row items-center">
        <Checkbox id="terms" {...args} />
        <Label htmlFor="terms">Accept terms and conditions</Label>
      </FormControl>
    );
  },
};
