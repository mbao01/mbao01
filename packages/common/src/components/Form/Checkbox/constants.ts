import { cva } from "../../../libs";

export const getCheckboxClasses = cva(
  "peer flex items-center justify-center h-4 w-4 shrink-0 rounded-xs border focus-visible:border-base-content focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-offset-1 focus-visible:ring-base-content disabled:cursor-not-allowed disabled:opacity-50",
  {
    variants: {
      variant: {
        primary:
          "border-primary data-[state=checked]:bg-primary data-[state=checked]:text-primary-content focus-visible:ring-primary",
        secondary:
          "border-secondary data-[state=checked]:bg-secondary data-[state=checked]:text-secondary-content focus-visible:ring-secondary",
        accent:
          "border-accent data-[state=checked]:bg-accent data-[state=checked]:text-accent-content focus-visible:ring-accent",
        success:
          "border-success data-[state=checked]:bg-success data-[state=checked]:text-success-content focus-visible:ring-success",
        warning:
          "border-warning data-[state=checked]:bg-warning data-[state=checked]:text-warning-content focus-visible:ring-warning",
        info: "border-info data-[state=checked]:bg-info data-[state=checked]:text-info-content focus-visible:ring-info",
        error:
          "border-error data-[state=checked]:bg-error data-[state=checked]:text-error-content focus-visible:ring-error",
      },
      size: {
        xs: "h-3 w-3",
        sm: "h-4 w-4",
        md: "h-5 w-5",
        lg: "h-7 w-7",
      },
      rounded: {
        xs: "rounded-xs",
        sm: "rounded-sm",
        md: "rounded-md",
        lg: "rounded-lg",
      },
    },
  }
);

export const getCheckboxIndicatorClasses = cva("flex items-center justify-center text-current");

export const getCheckboxIconClasses = cva("h-3.5 w-3.5", {
  variants: {
    size: {
      xs: "h-2.5 w-2.5",
      sm: "h-3 w-3",
      md: "h-4 w-4",
      lg: "h-6 w-6",
    },
  },
});
