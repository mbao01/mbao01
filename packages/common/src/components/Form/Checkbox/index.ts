export { Checkbox, CheckboxControlled } from "./Checkbox";
