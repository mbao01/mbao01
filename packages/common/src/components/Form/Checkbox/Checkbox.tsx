import * as React from "react";
import { type CheckboxProps } from "./types";
import { getCheckboxClasses } from "./constants";
import { cn } from "../../../utilities";

const Checkbox = React.forwardRef<HTMLInputElement, CheckboxProps>(
  ({ className, size, variant, ...props }, ref) => (
    <input
      ref={ref}
      type="checkbox"
      role="checkbox"
      className={cn(getCheckboxClasses({ size, variant }), className)}
      {...props}
    />
  )
);
Checkbox.displayName = "Checkbox";

export { Checkbox };
