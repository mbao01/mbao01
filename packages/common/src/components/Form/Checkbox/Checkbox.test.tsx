import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { describe, expect, it } from "vitest";
import { Checkbox } from "./Checkbox";

describe("Checkbox", () => {
  it("renders with the correct initial state", () => {
    const { asFragment } = render(<Checkbox name="username" defaultChecked />);

    expect(screen.getByRole("checkbox")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("handles user input correctly", async () => {
    const user = userEvent.setup();

    const { asFragment } = render(<Checkbox name="username" />);
    const checkbox = screen.getByRole("checkbox");

    await user.click(checkbox);

    expect(checkbox).toBeChecked();
    expect(asFragment()).toMatchSnapshot();
  });

  it("prevent user input when disabled", async () => {
    const user = userEvent.setup();

    const { asFragment } = render(<Checkbox name="username" disabled />);
    const checkbox = screen.getByRole("checkbox");

    await user.click(checkbox);

    expect(checkbox).toBeDisabled();
    expect(checkbox).not.toBeChecked();
    expect(asFragment()).toMatchSnapshot();
  });

  it.each(["primary", "secondary", "accent", "success", "warning", "info", "error"] as const)(
    "has %s variant",
    (variant) => {
      const { asFragment } = render(<Checkbox name="name" defaultChecked variant={variant} />);

      expect(screen.getByRole("checkbox")).toBeInTheDocument();
      expect(asFragment()).toMatchSnapshot();
    }
  );

  it.each([
    { size: "xs", name: "tiny" },
    { size: "sm", name: "small" },
    { size: "md", name: "base" },
    { size: "lg", name: "large" },
  ] as const)("has $name ($size)", ({ size, name }) => {
    const { asFragment } = render(<Checkbox name={name} size={size} />);

    expect(screen.getByRole("checkbox")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
