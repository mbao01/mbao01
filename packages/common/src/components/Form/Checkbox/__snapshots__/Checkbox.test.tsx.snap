// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Checkbox > handles user input correctly 1`] = `
<DocumentFragment>
  <button
    aria-checked="true"
    class="peer flex items-center justify-center h-4 w-4 shrink-0 rounded-xs border focus-visible:border-base-content focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-offset-1 focus-visible:ring-base-content disabled:cursor-not-allowed disabled:opacity-50"
    data-state="checked"
    role="checkbox"
    type="button"
    value="on"
  >
    <span
      class="flex items-center justify-center text-current"
      data-state="checked"
      style="pointer-events: none;"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-check h-3.5 w-3.5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20 6 9 17l-5-5"
        />
      </svg>
    </span>
  </button>
</DocumentFragment>
`;

exports[`Checkbox > has 'base' ('md') 1`] = `
<DocumentFragment>
  <button
    aria-checked="false"
    class="peer flex items-center justify-center shrink-0 rounded-xs border focus-visible:border-base-content focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-offset-1 focus-visible:ring-base-content disabled:cursor-not-allowed disabled:opacity-50 h-5 w-5"
    data-state="unchecked"
    role="checkbox"
    type="button"
    value="on"
  />
</DocumentFragment>
`;

exports[`Checkbox > has 'large' ('lg') 1`] = `
<DocumentFragment>
  <button
    aria-checked="false"
    class="peer flex items-center justify-center shrink-0 rounded-xs border focus-visible:border-base-content focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-offset-1 focus-visible:ring-base-content disabled:cursor-not-allowed disabled:opacity-50 h-7 w-7"
    data-state="unchecked"
    role="checkbox"
    type="button"
    value="on"
  />
</DocumentFragment>
`;

exports[`Checkbox > has 'small' ('sm') 1`] = `
<DocumentFragment>
  <button
    aria-checked="false"
    class="peer flex items-center justify-center shrink-0 rounded-xs border focus-visible:border-base-content focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-offset-1 focus-visible:ring-base-content disabled:cursor-not-allowed disabled:opacity-50 h-4 w-4"
    data-state="unchecked"
    role="checkbox"
    type="button"
    value="on"
  />
</DocumentFragment>
`;

exports[`Checkbox > has 'tiny' ('xs') 1`] = `
<DocumentFragment>
  <button
    aria-checked="false"
    class="peer flex items-center justify-center shrink-0 rounded-xs border focus-visible:border-base-content focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-offset-1 focus-visible:ring-base-content disabled:cursor-not-allowed disabled:opacity-50 h-3 w-3"
    data-state="unchecked"
    role="checkbox"
    type="button"
    value="on"
  />
</DocumentFragment>
`;

exports[`Checkbox > has accent variant 1`] = `
<DocumentFragment>
  <button
    aria-checked="true"
    class="peer flex items-center justify-center h-4 w-4 shrink-0 rounded-xs border focus-visible:border-base-content focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-offset-1 disabled:cursor-not-allowed disabled:opacity-50 border-accent data-[state=checked]:bg-accent data-[state=checked]:text-accent-content focus-visible:ring-accent"
    data-state="checked"
    role="checkbox"
    type="button"
    value="on"
  >
    <span
      class="flex items-center justify-center text-current"
      data-state="checked"
      style="pointer-events: none;"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-check h-3.5 w-3.5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20 6 9 17l-5-5"
        />
      </svg>
    </span>
  </button>
</DocumentFragment>
`;

exports[`Checkbox > has error variant 1`] = `
<DocumentFragment>
  <button
    aria-checked="true"
    class="peer flex items-center justify-center h-4 w-4 shrink-0 rounded-xs border focus-visible:border-base-content focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-offset-1 disabled:cursor-not-allowed disabled:opacity-50 border-error data-[state=checked]:bg-error data-[state=checked]:text-error-content focus-visible:ring-error"
    data-state="checked"
    role="checkbox"
    type="button"
    value="on"
  >
    <span
      class="flex items-center justify-center text-current"
      data-state="checked"
      style="pointer-events: none;"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-check h-3.5 w-3.5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20 6 9 17l-5-5"
        />
      </svg>
    </span>
  </button>
</DocumentFragment>
`;

exports[`Checkbox > has info variant 1`] = `
<DocumentFragment>
  <button
    aria-checked="true"
    class="peer flex items-center justify-center h-4 w-4 shrink-0 rounded-xs border focus-visible:border-base-content focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-offset-1 disabled:cursor-not-allowed disabled:opacity-50 border-info data-[state=checked]:bg-info data-[state=checked]:text-info-content focus-visible:ring-info"
    data-state="checked"
    role="checkbox"
    type="button"
    value="on"
  >
    <span
      class="flex items-center justify-center text-current"
      data-state="checked"
      style="pointer-events: none;"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-check h-3.5 w-3.5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20 6 9 17l-5-5"
        />
      </svg>
    </span>
  </button>
</DocumentFragment>
`;

exports[`Checkbox > has primary variant 1`] = `
<DocumentFragment>
  <button
    aria-checked="true"
    class="peer flex items-center justify-center h-4 w-4 shrink-0 rounded-xs border focus-visible:border-base-content focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-offset-1 disabled:cursor-not-allowed disabled:opacity-50 border-primary data-[state=checked]:bg-primary data-[state=checked]:text-primary-content focus-visible:ring-primary"
    data-state="checked"
    role="checkbox"
    type="button"
    value="on"
  >
    <span
      class="flex items-center justify-center text-current"
      data-state="checked"
      style="pointer-events: none;"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-check h-3.5 w-3.5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20 6 9 17l-5-5"
        />
      </svg>
    </span>
  </button>
</DocumentFragment>
`;

exports[`Checkbox > has rounded ('lg') 1`] = `
<DocumentFragment>
  <button
    aria-checked="false"
    class="peer flex items-center justify-center h-4 w-4 shrink-0 border focus-visible:border-base-content focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-offset-1 focus-visible:ring-base-content disabled:cursor-not-allowed disabled:opacity-50 rounded-lg"
    data-state="unchecked"
    role="checkbox"
    type="button"
    value="on"
  />
</DocumentFragment>
`;

exports[`Checkbox > has rounded ('md') 1`] = `
<DocumentFragment>
  <button
    aria-checked="false"
    class="peer flex items-center justify-center h-4 w-4 shrink-0 border focus-visible:border-base-content focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-offset-1 focus-visible:ring-base-content disabled:cursor-not-allowed disabled:opacity-50 rounded-md"
    data-state="unchecked"
    role="checkbox"
    type="button"
    value="on"
  />
</DocumentFragment>
`;

exports[`Checkbox > has rounded ('sm') 1`] = `
<DocumentFragment>
  <button
    aria-checked="false"
    class="peer flex items-center justify-center h-4 w-4 shrink-0 border focus-visible:border-base-content focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-offset-1 focus-visible:ring-base-content disabled:cursor-not-allowed disabled:opacity-50 rounded-sm"
    data-state="unchecked"
    role="checkbox"
    type="button"
    value="on"
  />
</DocumentFragment>
`;

exports[`Checkbox > has rounded ('xs') 1`] = `
<DocumentFragment>
  <button
    aria-checked="false"
    class="peer flex items-center justify-center h-4 w-4 shrink-0 border focus-visible:border-base-content focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-offset-1 focus-visible:ring-base-content disabled:cursor-not-allowed disabled:opacity-50 rounded-xs"
    data-state="unchecked"
    role="checkbox"
    type="button"
    value="on"
  />
</DocumentFragment>
`;

exports[`Checkbox > has secondary variant 1`] = `
<DocumentFragment>
  <button
    aria-checked="true"
    class="peer flex items-center justify-center h-4 w-4 shrink-0 rounded-xs border focus-visible:border-base-content focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-offset-1 disabled:cursor-not-allowed disabled:opacity-50 border-secondary data-[state=checked]:bg-secondary data-[state=checked]:text-secondary-content focus-visible:ring-secondary"
    data-state="checked"
    role="checkbox"
    type="button"
    value="on"
  >
    <span
      class="flex items-center justify-center text-current"
      data-state="checked"
      style="pointer-events: none;"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-check h-3.5 w-3.5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20 6 9 17l-5-5"
        />
      </svg>
    </span>
  </button>
</DocumentFragment>
`;

exports[`Checkbox > has success variant 1`] = `
<DocumentFragment>
  <button
    aria-checked="true"
    class="peer flex items-center justify-center h-4 w-4 shrink-0 rounded-xs border focus-visible:border-base-content focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-offset-1 disabled:cursor-not-allowed disabled:opacity-50 border-success data-[state=checked]:bg-success data-[state=checked]:text-success-content focus-visible:ring-success"
    data-state="checked"
    role="checkbox"
    type="button"
    value="on"
  >
    <span
      class="flex items-center justify-center text-current"
      data-state="checked"
      style="pointer-events: none;"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-check h-3.5 w-3.5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20 6 9 17l-5-5"
        />
      </svg>
    </span>
  </button>
</DocumentFragment>
`;

exports[`Checkbox > has warning variant 1`] = `
<DocumentFragment>
  <button
    aria-checked="true"
    class="peer flex items-center justify-center h-4 w-4 shrink-0 rounded-xs border focus-visible:border-base-content focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-offset-1 disabled:cursor-not-allowed disabled:opacity-50 border-warning data-[state=checked]:bg-warning data-[state=checked]:text-warning-content focus-visible:ring-warning"
    data-state="checked"
    role="checkbox"
    type="button"
    value="on"
  >
    <span
      class="flex items-center justify-center text-current"
      data-state="checked"
      style="pointer-events: none;"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-check h-3.5 w-3.5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20 6 9 17l-5-5"
        />
      </svg>
    </span>
  </button>
</DocumentFragment>
`;

exports[`Checkbox > prevent user input when disabled 1`] = `
<DocumentFragment>
  <button
    aria-checked="false"
    class="peer flex items-center justify-center h-4 w-4 shrink-0 rounded-xs border focus-visible:border-base-content focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-offset-1 focus-visible:ring-base-content disabled:cursor-not-allowed disabled:opacity-50"
    data-disabled=""
    data-state="unchecked"
    disabled=""
    role="checkbox"
    type="button"
    value="on"
  />
</DocumentFragment>
`;

exports[`Checkbox > renders with an indeterminate state 1`] = `
<DocumentFragment>
  <button
    aria-checked="mixed"
    class="peer flex items-center justify-center h-4 w-4 shrink-0 rounded-xs border focus-visible:border-base-content focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-offset-1 focus-visible:ring-base-content disabled:cursor-not-allowed disabled:opacity-50"
    data-state="indeterminate"
    role="checkbox"
    type="button"
    value="on"
  >
    <span
      class="flex items-center justify-center text-current"
      data-state="indeterminate"
      style="pointer-events: none;"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-minus h-3.5 w-3.5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 12h14"
        />
      </svg>
    </span>
  </button>
</DocumentFragment>
`;

exports[`Checkbox > renders with the correct initial state 1`] = `
<DocumentFragment>
  <button
    aria-checked="true"
    class="peer flex items-center justify-center h-4 w-4 shrink-0 rounded-xs border focus-visible:border-base-content focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-offset-1 focus-visible:ring-base-content disabled:cursor-not-allowed disabled:opacity-50"
    data-state="checked"
    role="checkbox"
    type="button"
    value="on"
  >
    <span
      class="flex items-center justify-center text-current"
      data-state="checked"
      style="pointer-events: none;"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-check h-3.5 w-3.5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20 6 9 17l-5-5"
        />
      </svg>
    </span>
  </button>
</DocumentFragment>
`;
